### ISSUE TITLE
Split vscode NPM node module into @types/vscode and vscode-test

### ISSUE THREAD

**Issue by octref** [Microsoft/vscode#69668]
Per the recent incidents (https://github.com/Microsoft/vscode/issues/69656), I think it's a good opportunity for us to split `vscode` node modules into 2 pieces:

- `@types/vscode`
- `vscode-test`

### Explanation

`vscode` node module contains 3 scripts:

- `compile`: no longer used
- `install`: used to fetch `vscode.d.ts` according to `engine.vscode` in `package.json`
- `test`: used for extension test (also `lib/testRunner`)

### Problems

- `install` asks update server for finding matching `vscode.d.ts` version (actual dts files fetched from GitHub). When update service is down, all extension deps install (CI or local development) could fail.
- `yarn install` and `npm install` can't be completed offline, even if they have caching: https://github.com/Microsoft/vscode-extension-vscode/issues/93
- `test` depends on a lot of node modules (13MB). Everyone has to download them even if they just want to use the API.
- `test` needs some improvements. For example, for running extension tests against multiple workspaces, I had to pull out the test scripts and modify them manually [Vetur](https://github.com/vuejs/vetur/blob/master/test/codeTestRunner.ts). Also https://github.com/Microsoft/vscode-extension-vscode/issues/124, people want functionalities exposed from `test` so they can build upon them.

### Benefits of splitting `vscode` into `@types/vscode` and `vscode-test`

- No awkward postinstall scripts after each `npm install` for extension
- Users don't have to pull in the 11MB dependencies of `test`. They pull a few kb of `vscode.d.ts` from `@types/vscode`, and it's fully cached by npm/yarn.
- `vscode.d.ts` are hosted on NPM which is pretty reliable. `d.ts` live in a `@types` package as they should be.
- `test` now lives in its own repo. It's easier to make improvements, release new versions, expose functionalities (for example, `import { downloadVSCodeExecutable } from vscode-test`).

### Challenges

We can't put all versions of `vscode.d.ts` into `@types/vscode` and let TS infer which one, from `engines.vscode`.

However I think we can publish `vscode.d.ts` the same version as we do for VS Code, since all our 1.x API are backwards compatible. Most users also auto-update to latest version, so extension authors could use `"@types/vscode": "^1.31.0"` and safely get minor updates.

/cc @Microsoft/vscode for thoughts.

**Comment by alexandrudima** [comment]
:+1: I like the proposal, I would suggest the small twist of shipping the `.d.ts` file directly within the `vscode` npm module instead of using `@types/vscode`. But that means we have to remember to always publish a new `vscode` npm module when we create a stable release. That should be fine, given we do it strictly after a VS Code release.

I would also vote for rewriting the script which downloads a vscode release to use `https` directly and the native installed `tar` instead of 13MB of `gulp-*` dependencies.

**Comment by octref** [comment]
> rewriting the script which downloads a vscode release to use https directly

Agreed, something like this already works for macOS/Linux, which is what I'm using on Travis. `vscode-test` should also contain minimal or zero dependencies, so we'd be exempted from issues like `event-stream`.

```js
let downloadPlatform

switch (process.platform) {
  case 'darwin':
    downloadPlatform = 'darwin'
    break
  case 'win32':
    downloadPlatform = 'win32-archive'
    break
  default:
    downloadPlatform = 'linux-x64'
}

const url = `https://update.code.visualstudio.com/latest/${downloadPlatform}/stable`

const https = require('https')
const fs = require('fs')
const cp = require('child_process')

https.get(url, res => {
  const zipUrl = res.headers.location

  if (zipUrl.endsWith('.zip')) {
    const outStream = fs.createWriteStream('vscode.zip')
    https.get(zipUrl, res => {
      res.pipe(outStream)
    })
    outStream.on('close', () => {
      if (process.platform === 'win32') {
        // need some windows powershell script or a zip library that supports windows
      } else {
        cp.execSync('unzip vscode.zip')
      }
    })
  } else {
    const outStream = fs.createWriteStream('vscode.tgz')
    https.get(zipUrl, res => {
      res.pipe(outStream)
    })
    outStream.on('close', () => {
      cp.execSync('tar -xzf vscode.tgz')
    })
  }
})
```
